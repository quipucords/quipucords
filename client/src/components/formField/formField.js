import React from 'react';
import PropTypes from 'prop-types';
import { Form, Grid } from 'patternfly-react';
import helpers from '../../common/helpers';

const FormField = ({
  children,
  colLabel,
  colLabelClassName,
  colField,
  colFieldClassName,
  error,
  errorMessage,
  id,
  label,
  ...props
}) => {
  const setId = id || helpers.generateId();

  return (
    <Form.FormGroup controlId={setId} validationState={error ? 'error' : null} {...props}>
      <Grid.Col componentClass={Form.ControlLabel} className={colLabelClassName} sm={colLabel}>
        {label}
      </Grid.Col>
      <Grid.Col className={colFieldClassName} sm={colField}>
        {children}
        {error && <Form.HelpBlock>{errorMessage}</Form.HelpBlock>}
      </Grid.Col>
    </Form.FormGroup>
  );
};

const doesntHaveMinimumCharacters = (value, characters = 5) => typeof value === 'string' && value.length < characters;
const isEmpty = value => !value || value === '';

const fieldValidation = {
  doesntHaveMinimumCharacters,
  isEmpty
};

FormField.propTypes = {
  children: PropTypes.node.isRequired,
  colLabel: PropTypes.number,
  colLabelClassName: PropTypes.string,
  colField: PropTypes.number,
  colFieldClassName: PropTypes.string,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  errorMessage: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.node
};

FormField.defaultProps = {
  colLabel: 3,
  colLabelClassName: null,
  colField: 9,
  colFieldClassName: null,
  error: null,
  errorMessage: null,
  id: null,
  label: null
};

export { FormField as default, FormField, fieldValidation };
