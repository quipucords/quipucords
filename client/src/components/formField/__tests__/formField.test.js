import React from 'react';
import { mount } from 'enzyme';
import { FormField, fieldValidation } from '../formField';

describe('FormField Component', () => {
  it('should render', () => {
    const props = { id: 'test' };

    const component = mount(
      <FormField {...props}>
        <input id="test" type="text" value="" readOnly />
      </FormField>
    );

    expect(component.render()).toMatchSnapshot();
    expect(
      component
        .find('label')
        .at(0)
        .render()
    ).toMatchSnapshot();
  });

  it('should have isEmpty validation', () => {
    expect(fieldValidation.isEmpty('')).toEqual(true);
    expect(fieldValidation.isEmpty('test')).toEqual(false);
  });

  it('should have doesntHaveMinimumCharacters validation', () => {
    expect(fieldValidation.doesntHaveMinimumCharacters('', 5)).toEqual(true);
    expect(fieldValidation.doesntHaveMinimumCharacters('test test', 5)).toEqual(false);
  });
});
