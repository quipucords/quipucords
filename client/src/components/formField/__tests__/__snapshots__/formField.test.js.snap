// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField Component should render 1`] = `
<div
  class="form-group"
>
  <label
    class="col-sm-3 control-label"
    for="test"
  />
  <div
    class="col-sm-9"
  >
    <input
      id="test"
      readonly=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`FormField Component should render 2`] = `
<label
  class="col-sm-3 control-label"
  for="test"
/>
`;
